### PR TITLE
Mount build source

### PR DESCRIPTION
**Pull request by lucasdemarchi**

### CODE REVIEW

File: mkosi [hunk 1]
```diff
@@ -2605,10 +2607,9 @@ def parse_args() -> CommandLineArguments:
     group.add_argument("--build-package", action=CommaDelimitedListAction, dest='build_packages', default=[], help='Additional packages needed for build script', metavar='PACKAGE')
     group.add_argument("--postinst-script", help='Postinstall script to run inside image', metavar='PATH')
     group.add_argument("--finalize-script", help='Postinstall script to run outside image', metavar='PATH')
-    group.add_argument('--use-git-files', type=parse_boolean,
-                       help='Ignore any files that git itself ignores (default: guess)')
+    group.add_argument("--copy-sources-method", choices=('plain', 'git'), help="Method used to copy build sources to the build image. 'plain': normal file copy; 'git': use git-ls-files, ignoring any file that git itself ignores. (default: guess)")
     group.add_argument('--git-files', choices=('cached', 'others'),
-                       help='Whether to include untracked files (default: others)')
+                       help='Whether to include untracked files, used together with --copy-sources-method=git (default: others)')
```
Comment (poettering) [inline mkosi]: if you clean this up with --copy-sources-method, maybe we can fold this in too, i.e. introduce --copy-sources-method=git-cached vs. --copy-sources-method=git-others?

I think fewer options are nice, in particular if they can be safely flattened into one enum.

File: mkosi [hunk 2]
```diff
@@ -2920,6 +2921,11 @@ def process_setting(args: CommandLineArguments, section: str, key: str, value: A
         elif key == "BuildSources":
             if args.build_sources is None:
                 args.build_sources = value
+        elif key == "CopySourcesMethod":
```
Comment (poettering) [inline mkosi]: maybe name this "SourceFileFiltering=" or so?

Comment (poettering) [reply mkosi]: or SourceFileAccess=? SourceFileTransfer=?

And then name the options `copy-all`, `copy-git-cached`, `copy-git-others`, `mount`?

**Comment by poettering** [comment]
looks great otherwise

**Comment by poettering** [comment]
(also needs a rebase)

**Comment by poettering** [comment]
Looks good to me, but I think @keszybz should have a look at this, as he's a better pythonista then me, and this looks like something which would benefit from a pythonic eye... @keszybz?

**Comment by poettering** [comment]
(btw, unrelated to this, I think we should tag v5 soon, after maybe fixing #253. or does anyone have more stuff they'd like to see merged before?)

**Comment by lucasdemarchi** [comment]
One thing I'd want to do is to take a look on merging the config and arg parsing... It's annoying me how error prone and verbose it is right now. However this is more an internal organization change, not a real feature. There's also what we discussed in #172, but that can wait. I'm perfectly fine with a new release.

**Comment by poettering** [comment]
Hmm, something is not right with this one... mkosi now wants to copy my whole src tree all the time, and doesn't honour the git stuff anymore... This should not have changed the default logic really

**Comment by poettering** [comment]
Traceback (most recent call last):
  File "../../mkosi/mkosi", line 301, in copy_fd
    _reflink(oldfd, newfd)
  File "../../mkosi/mkosi", line 297, in _reflink
    fcntl.ioctl(newfd, FICLONE, oldfd)
OSError: [Errno 18] Invalid cross-device link

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "../../mkosi/mkosi", line 4055, in <module>
    main()
  File "../../mkosi/mkosi", line 4045, in main
    build_stuff(args)
  File "../../mkosi/mkosi", line 3878, in build_stuff
    raw, tar, root_hash = build_image(args, workspace, run_build_script=True)
  File "../../mkosi/mkosi", line 3732, in build_image
    install_build_src(args, workspace.name, run_build_script, for_cache)
  File "../../mkosi/mkosi", line 2094, in install_build_src
    copy_git_files(args.build_sources, target, source_file_transfer=source_file_transfer)
  File "../../mkosi/mkosi", line 2071, in copy_git_files
    copy_file(src_path, dest_path)
  File "../../mkosi/mkosi", line 330, in copy_file
    copy_fd(oldfd, newfd)
  File "../../mkosi/mkosi", line 306, in copy_fd
    open(newfd, 'wb', closefd=False))
  File "/usr/lib64/python3.7/shutil.py", line 82, in copyfileobj
    fdst.write(buf)
OSError: [Errno 28] No space left on device

**Comment by poettering** [comment]
hmm, it appears the behaviour changed from using "git-copy-cached" as before... any reason for that, or is that accidental?

**Comment by lucasdemarchi** [comment]
```diff
diff --git a/mkosi b/mkosi
index 9e13c3a..8711d75 100755
--- a/mkosi
+++ b/mkosi
@@ -2079,7 +2079,7 @@ def install_build_src(args, workspace, run_build_script, for_cache):
 
             source_file_transfer = args.source_file_transfer
             if source_file_transfer is None and (os.path.exists('.git') or os.path.exists(os.path.join(args.build_sources, '.git'))):
-                source_file_transfer = SourceFileTransfer.copy_git_others
+                source_file_transfer = SourceFileTransfer.copy_git_cached
 
             if source_file_transfer in (SourceFileTransfer.copy_git_others, SourceFileTransfer.copy_git_cached):
                 copy_git_files(args.build_sources, target, source_file_transfer=source_file_transfer)
@@ -2707,7 +2707,7 @@ def parse_args() -> CommandLineArguments:
     group.add_argument("--finalize-script", help='Postinstall script to run outside image', metavar='PATH')
     group.add_argument("--source-file-transfer", type=SourceFileTransfer, choices=list(SourceFileTransfer), default=None,
                        help="Method used to copy build sources to the build image." +
-                       "; ".join([f"'{k}': {v}" for k, v in SourceFileTransfer.doc().items()]) + " (default: copy-git-others if in a git repository, otherwise copy-all)")
+                       "; ".join([f"'{k}': {v}" for k, v in SourceFileTransfer.doc().items()]) + " (default: copy-git-cached if in a git repository, otherwise copy-all)")
     group.add_argument("--with-network", action='store_true', help='Run build and postinst scripts with network access (instead of private network)', default=None)
     group.add_argument("--settings", dest='nspawn_settings', help='Add in .spawn settings file', metavar='PATH')
 
```

**Pull request merged**